### PR TITLE
Cuda fixes for tests

### PR DESCRIPTION
**Pull request by agibsonccc**
Decompress cuda ndarray when it exists
Misc samediff fixes with execution
Add missing cuda blas delegator resource class (basically no op) Fix null pointer in atomic allocator with empty arrays Disable paragraph vectors tests on cuda (takes too long and isn't really intended to be used with cuda anyways) Add back missing declarations for migration from old optimizers

## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [ X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

### CODE REVIEW

**Review by treo — DISMISSED** [review]

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/workspace/WorkspaceUtils.java [hunk 1]
```diff
@@ -152,7 +153,41 @@ public static int getShapeBufferRequireMemoryForWorkspace(INDArray arr) {
...
+     * etAligned(arr.length() * arr.dataType().width())
+     *
+     * This is due to shape buffers from cuda only being allocated
+     * from a cache rather than workspaces itself.
+     * @param arr the array to get the required memory for.
+     * @return
+     */
     public static int getTotalRequiredMemoryForWorkspace(INDArray arr) {
-        return getAligned(arr.length() * arr.dataType().width()) + getAligned(arr.shapeInfoJava().length * DataType.INT64.width());
+        if(!Nd4j.getBackend().getNDArrayClass().getName().toLowerCase().contains("cu")) {
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/workspace/WorkspaceUtils.java:174]: This looks more like a quick hack that can break in unexpected ways.

Also you are missing a `g` in the documentation:
```
     * GPUS will only be:
     * etAligned(arr.length() * arr.dataType().width())
```

File: platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/linalg/workspace/DebugModeTests.java [hunk 2]
```diff
@@ -134,6 +133,7 @@ public void testSpillMode_2(Nd4jBackend backend) {
             assertEquals(0, ws.getDeviceOffset());
 
             // array buffer should be spilled now
+            int shapeBufferAdd = backend.getClass().getName().toLowerCase().contains("cu") ? 0 : WorkspaceUtils.getShapeBufferRequireMemoryForWorkspace(array);
```
Comment (treo) [inline platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/linalg/workspace/DebugModeTests.java]: Same note about the hack applies here too. 

If you need it in multiple places, put it into a helper method.

**Review by treo — CHANGES_REQUESTED** [review]

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/workspace/WorkspaceUtils.java (hunk 1 above)
Comment (agibsonccc) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/workspace/WorkspaceUtils.java:174]: Fixed! I added new metric recording and then use those in the unit tests. There's some hacks that can't be avoided due to the "answer" being different depending on the backend but it's at least checked against the expected metrics.

File: platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/linalg/workspace/DebugModeTests.java (hunk 2 above)
Comment (agibsonccc) [reply platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/linalg/workspace/DebugModeTests.java]: Same answer as above.

**Review by treo — APPROVED** [review]

**Pull request merged**